source/common/grpc/google_grpc_utils.cc

```diff
@@ -62,12 +62,17 @@ grpc::ByteBuffer GoogleGrpcUtils::makeByteBuffer(Buffer::InstancePtr&& buffer_in
   return {&slices[0], slices.size()};
 }
 
-struct ByteBufferContainer {
-  ByteBufferContainer(int ref_count) : ref_count_(ref_count) {}
-  ~ByteBufferContainer() { ::free(fragments_); }
-  uint32_t ref_count_;
-  Buffer::BufferFragmentImpl* fragments_ = nullptr;
-  std::vector<grpc::Slice> slices_;
+class GrpcSliceBufferFragmentImpl : public Buffer::BufferFragment {
+public:
+  explicit GrpcSliceBufferFragmentImpl(grpc::Slice&& slice) : slice_(std::move(slice)) {}
+
+  // Buffer::BufferFragment
+  const void* data() const override { return slice_.begin(); }
+  size_t size() const override { return slice_.size(); }
+  void done() override { delete this; }
```

Comment:
I am not sure if this is the best behavior: It assumes `GrpcSliceBufferFragmentImpl` is always on heap, which is not necessary although the only usage is :)

Reply:
I would choose to release the `slice_` here and don't allocate `GrpcSliceBufferFragmentImpl`.
I don't see the best way to release the slice but may be `slice_ = grpc::Slice()` is the closest expression.

Reply:
That doesn't work, `addBufferFragment` only takes a reference but not ownership, so it need to being deleted here. This is no different from `Buffer::BufferFragmentImpl` with `delete frag;` in previous commit.

Reply:
Learnt a lot from the discussion before I write any code to crash the envoy :)